examples/resnet-concurrent.cpp

```diff
@@ -0,0 +1,219 @@
+                      Placeholder *output, std::unique_ptr<Context> ctx,
+                      std::atomic<size_t> &returned,
+                      std::promise<void> &finished) {
+  device->runFunction(
+      "resnet50", std::move(ctx),
+      [id, path, output, &returned, &finished](RunIdentifierTy, ResultCode r,
+                                               std::unique_ptr<Context> ctx) {
+        size_t maxIdx = ctx->get(output)->getHandle<>().minMaxArg().second;
+
+        llvm::outs() << "(" << id << ") " << path << ": " << maxIdx << "\n";
+        if (++returned == maxImages) {
+          finished.set_value();
+        }
+      });
+}
+
+/// Run ResNet concurrently on a fixed number of CPU Devices
+int main(int argc, char **argv) {
+  llvm::cl::ParseCommandLineOptions(
+      argc, argv, "Run ResNet concurrently on a fixed number of CPU devices");
+
+  llvm::outs() << "Initializing " << numDevices << " CPU Devices.\n";
+  std::vector<std::unique_ptr<CPUDeviceManager>> devices;
+  for (unsigned int i = 0; i < numDevices; ++i) {
+    devices.emplace_back(std::make_unique<CPUDeviceManager>());
+    devices[i]->init();
+  }
+
+  /// Load and compile model.
+
+  Module module;
+  TypeRef inputType(module.uniqueType(ElemKind::FloatTy, {1, 3, 224, 224}));
+  Placeholder *input, *output;
+
+  std::tie(input, output) = loadResnet50Model(inputType, module);
+  auto compiledFunction = compileModel(module);
+
+  FunctionMapTy functions;
+  functions.emplace("resnet50", compiledFunction.get());
+
+  std::vector<std::future<ResultCode>> compiles;
+  compiles.reserve(numDevices);
+
+  for (unsigned int i = 0; i < numDevices; ++i) {
+    compiles.push_back(addToDevice(i, devices[i].get(), module, functions));
+  }
+
+  for (auto &f : compiles) {
+    f.wait_for(30s);
+    ResultCode code = f.get();
+    if (code != ResultCode::Ready) {
+      return 1;
+    }
+  }
+
+  llvm::outs() << "Loading files from " << inputDirectory << "\n";
+  std::error_code code;
+  llvm::sys::fs::directory_iterator dirIt(inputDirectory, code);
+  if (code.value()) {
+    llvm::errs() << "Couldn't read from directory: " << inputDirectory
+                 << " - code" << code.value() << "\n";
+    exit(code.value());
+  }
+
+  std::promise<void> finished;
+
+  size_t started = 0;
+  std::atomic<size_t> returned{0};
+
+  // Run up to maxImages classifications.
+
+  while (started++ < maxImages) {
+    if (code.value() != 0 || dirIt == llvm::sys::fs::directory_iterator()) {
+      started--;
+      returned += maxImages - started;
+
+      if (returned == maxImages) {
+        finished.set_value();
+      }
+      break;
+    }
+
+    std::string path = dirIt->path();
+
+    auto image =
+        readPngImageAndPreprocess(path, ImageNormalizationMode::k0to1,
+                                  ImageChannelOrder::BGR, ImageLayout::NCHW,
+                                  /* useImagenetNormalization */ true);
+
+    Tensor batch(inputType);
```

Comment:
Why can't you just pass `image` to `updateInputPlaceholders()`?

Reply:
I'm actually not sure, I took this pattern from the resnet-verify example. You're right that you don't need it though.